src/libraries/System.Net.Sockets/tests/FunctionalTests/ReceiveMessageFrom.cs

```diff
@@ -3,173 +3,124 @@
         {
-            if (Socket.OSSupportsIPv6)
-            {
-                using (Socket receiver = new Socket(AddressFamily.InterNetworkV6, SocketType.Dgram, ProtocolType.Udp))
-                {
-                    int port = receiver.BindToAnonymousPort(IPAddress.IPv6Loopback);
-                    receiver.SetSocketOption(SocketOptionLevel.IPv6, SocketOptionName.PacketInformation, true);
+            const int DatagramSize = 256;
+            const int DatagramsToSend = 16;
 
-                    receiver.ForceNonBlocking(forceNonBlocking);
+            IPAddress address = ipv4 ? IPAddress.Loopback : IPAddress.IPv6Loopback;
+            using Socket receiver = new Socket(address.AddressFamily, SocketType.Dgram, ProtocolType.Udp);
+            using Socket sender = new Socket(address.AddressFamily, SocketType.Dgram, ProtocolType.Udp);
 
-                    Socket sender = new Socket(AddressFamily.InterNetworkV6, SocketType.Dgram, ProtocolType.Udp);
-                    sender.Bind(new IPEndPoint(IPAddress.IPv6Loopback, 0));
+            receiver.SetSocketOption(ipv4 ? SocketOptionLevel.IP : SocketOptionLevel.IPv6, SocketOptionName.PacketInformation, true);
+            ConfigureNonBlocking(sender);
+            ConfigureNonBlocking(receiver);
 
-                    sender.ForceNonBlocking(forceNonBlocking);
+            receiver.BindToAnonymousPort(address);
+            sender.BindToAnonymousPort(address);
 
-                    sender.SendTo(new byte[1024], new IPEndPoint(IPAddress.IPv6Loopback, port));
+            byte[] sendBuffer = new byte[DatagramSize];
+            byte[] receiveBuffer = new byte[DatagramSize];
+            Random rnd = new Random(0);
 
-                    IPPacketInformation packetInformation;
-                    SocketFlags flags = SocketFlags.None;
-                    EndPoint remoteEP = new IPEndPoint(IPAddress.IPv6Any, 0);
+            IPEndPoint remoteEp = new IPEndPoint(ipv4 ? IPAddress.Any : IPAddress.IPv6Any, 0);
 
-                    int len = receiver.ReceiveMessageFrom(new byte[1024], 0, 1024, ref flags, ref remoteEP, out packetInformation);
+            for (int i = 0; i < DatagramsToSend; i++)
+            {
+                rnd.NextBytes(sendBuffer);
+                sender.SendTo(sendBuffer, receiver.LocalEndPoint);
 
-                    Assert.Equal(1024, len);
-                    Assert.Equal(sender.LocalEndPoint, remoteEP);
-                    Assert.Equal(((IPEndPoint)sender.LocalEndPoint).Address, packetInformation.Address);
+                SocketReceiveMessageFromResult result = await ReceiveMessageFromAsync(receiver, receiveBuffer, remoteEp);
+                IPPacketInformation packetInformation = result.PacketInformation;
 
-                    sender.Dispose();
-                }
+                Assert.Equal(DatagramSize, result.ReceivedBytes);
+                Assert.Equal(Fletcher32.Checksum(sendBuffer), Fletcher32.Checksum(receiveBuffer));
```

Comment:
Nit: Instead of comparing checksums, why not just compare the data itself? It's not like it's that large. We added some helpers recently to ensure that comparing byte arrays/spans is efficient.

See here: https://github.com/dotnet/runtime/blob/master/src/libraries/Common/tests/TestUtilities/System/AssertExtensions.cs#L441